File: python/llm_client.py
"""
LLM Client for Prompt Engineering Trading

Provides a unified interface for interacting with various LLM APIs
including OpenAI, Anthropic, and local models via Ollama.
"""

import json
import os
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Dict, List, Optional, Any
import asyncio

try:
    import httpx
except ImportError:
    httpx = None

try:
    import aiohttp
except ImportError:
    aiohttp = None


@dataclass
class LLMResponse:
    """Response from LLM API."""
    content: str
    model: str
    usage: Optional[Dict[str, int]] = None
    finish_reason: Optional[str] = None


class BaseLLMClient(ABC):
    """Abstract base class for LLM clients."""

    @abstractmethod
    async def complete(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> str:
        """Generate completion for the given prompt."""
        pass

    @abstractmethod
    async def complete_with_metadata(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> LLMResponse:
        """Generate completion with full metadata."""
        pass


class OpenAIClient(BaseLLMClient):
    """
    OpenAI API client.

    Supports GPT-4 and GPT-3.5-turbo models.
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = "gpt-4",
        base_url: str = "https://api.openai.com/v1"
    ):
        self.api_key = api_key or os.getenv("OPENAI_API_KEY")
        self.model = model
        self.base_url = base_url

        if not self.api_key:
            raise ValueError("OpenAI API key is required")

    async def complete(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> str:
        """Generate completion."""
        response = await self.complete_with_metadata(
            prompt, temperature, max_tokens, **kwargs
        )
        return response.content

    async def complete_with_metadata(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> LLMResponse:
        """Generate completion with metadata."""
        if httpx is None:
            raise ImportError("httpx is required for OpenAI client")

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }

        data = {
            "model": self.model,
            "messages": [{"role": "user", "content": prompt}],
            "temperature": temperature,
            "max_tokens": max_tokens,
            **kwargs
        }

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/chat/completions",
                headers=headers,
                json=data,
                timeout=60.0
            )
            response.raise_for_status()
            result = response.json()

        choice = result["choices"][0]
        return LLMResponse(
            content=choice["message"]["content"],
            model=result["model"],
            usage=result.get("usage"),
            finish_reason=choice.get("finish_reason")
        )


class AnthropicClient(BaseLLMClient):
    """
    Anthropic API client.

    Supports Claude models.
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = "claude-3-sonnet-20240229",
        base_url: str = "https://api.anthropic.com/v1"
    ):
        self.api_key = api_key or os.getenv("ANTHROPIC_API_KEY")
        self.model = model
        self.base_url = base_url

        if not self.api_key:
            raise ValueError("Anthropic API key is required")

    async def complete(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> str:
        """Generate completion."""
        response = await self.complete_with_metadata(
            prompt, temperature, max_tokens, **kwargs
        )
        return response.content

    async def complete_with_metadata(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> LLMResponse:
        """Generate completion with metadata."""
        if httpx is None:
            raise ImportError("httpx is required for Anthropic client")

        headers = {
            "x-api-key": self.api_key,
            "Content-Type": "application/json",
            "anthropic-version": "2023-06-01"
        }

        data = {
            "model": self.model,
            "messages": [{"role": "user", "content": prompt}],
            "temperature": temperature,
            "max_tokens": max_tokens,
            **kwargs
        }

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/messages",
                headers=headers,
                json=data,
                timeout=60.0
            )
            response.raise_for_status()
            result = response.json()

        return LLMResponse(
            content=result["content"][0]["text"],
            model=result["model"],
            usage=result.get("usage"),
            finish_reason=result.get("stop_reason")
        )


class OllamaClient(BaseLLMClient):
    """
    Ollama client for local LLM models.

    Supports any model available in Ollama.
    """

    def __init__(
        self,
        model: str = "llama2",
        base_url: str = "http://localhost:11434"
    ):
        self.model = model
        self.base_url = base_url

    async def complete(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> str:
        """Generate completion."""
        response = await self.complete_with_metadata(
            prompt, temperature, max_tokens, **kwargs
        )
        return response.content

    async def complete_with_metadata(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> LLMResponse:
        """Generate completion with metadata."""
        if httpx is None:
            raise ImportError("httpx is required for Ollama client")

        data = {
            "model": self.model,
            "prompt": prompt,
            "stream": False,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,
                **kwargs
            }
        }

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/api/generate",
                json=data,
                timeout=120.0
            )
            response.raise_for_status()
            result = response.json()

        return LLMResponse(
            content=result["response"],
            model=result["model"],
            usage=None,
            finish_reason="stop" if result.get("done") else None
        )


class MockLLMClient(BaseLLMClient):
    """
    Mock LLM client for testing and demos.

    Returns predefined responses based on prompt content.
    """

    def __init__(self, responses: Optional[Dict[str, str]] = None):
        self.responses = responses or {}
        self.default_response = json.dumps({
            "sentiment": "NEUTRAL",
            "confidence": 50,
            "signal": "HOLD",
            "reasoning": "Mock response for testing"
        })

    async def complete(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> str:
        """Generate mock completion."""
        # Check for matching response
        for key, response in self.responses.items():
            if key.lower() in prompt.lower():
                return response

        return self.default_response

    async def complete_with_metadata(
        self,
        prompt: str,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        **kwargs
    ) -> LLMResponse:
        """Generate mock completion with metadata."""
        content = await self.complete(prompt, temperature, max_tokens, **kwargs)
        return LLMResponse(
            content=content,
            model="mock-model",
            usage={"prompt_tokens": len(prompt.split()), "completion_tokens": len(content.split())},
            finish_reason="stop"
        )


def create_llm_client(
    provider: str = "mock",
    **kwargs
) -> BaseLLMClient:
    """
    Factory function to create LLM client.

    Args:
        provider: LLM provider ("openai", "anthropic", "ollama", "mock")
        **kwargs: Provider-specific arguments

    Returns:
        Configured LLM client
    """
    providers = {
        "openai": OpenAIClient,
        "anthropic": AnthropicClient,
        "ollama": OllamaClient,
        "mock": MockLLMClient
    }

    if provider not in providers:
        raise ValueError(f"Unknown provider: {provider}. Choose from: {list(providers.keys())}")

    return providers[provider](**kwargs)

File: rust/src/llm_client.rs
//! LLM Client abstraction for multiple providers.

use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::error::{TradingError, TradingResult};

/// Configuration for LLM requests.
#[derive(Debug, Clone)]
pub struct LLMConfig {
    pub temperature: f32,
    pub max_tokens: u32,
    pub model: String,
}

impl Default for LLMConfig {
    fn default() -> Self {
        Self {
            temperature: 0.3,
            max_tokens: 1000,
            model: "gpt-4".to_string(),
        }
    }
}

/// Trait for LLM clients.
#[async_trait]
pub trait LLMClient: Send + Sync {
    /// Complete a prompt and return the response.
    async fn complete(&self, prompt: &str, config: &LLMConfig) -> TradingResult<String>;

    /// Get the provider name.
    fn provider(&self) -> &str;
}

/// Mock LLM client for testing.
pub struct MockLLMClient {
    responses: HashMap<String, String>,
    default_response: String,
}

impl MockLLMClient {
    /// Create a new mock client with default response.
    pub fn new() -> Self {
        Self {
            responses: HashMap::new(),
            default_response: r#"{"sentiment": "NEUTRAL", "confidence": 50}"#.to_string(),
        }
    }

    /// Create a mock client with specific responses.
    pub fn with_responses(responses: HashMap<String, String>) -> Self {
        Self {
            responses,
            default_response: r#"{"sentiment": "NEUTRAL", "confidence": 50}"#.to_string(),
        }
    }

    /// Set the default response.
    pub fn set_default_response(&mut self, response: String) {
        self.default_response = response;
    }

    /// Add a response for a specific key.
    pub fn add_response(&mut self, key: &str, response: String) {
        self.responses.insert(key.to_string(), response);
    }
}

impl Default for MockLLMClient {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl LLMClient for MockLLMClient {
    async fn complete(&self, prompt: &str, _config: &LLMConfig) -> TradingResult<String> {
        // Try to match prompt to a response key
        for (key, response) in &self.responses {
            if prompt.to_lowercase().contains(&key.to_lowercase()) {
                return Ok(response.clone());
            }
        }
        Ok(self.default_response.clone())
    }

    fn provider(&self) -> &str {
        "mock"
    }
}

/// OpenAI API client.
pub struct OpenAIClient {
    api_key: String,
    base_url: String,
    client: reqwest::Client,
}

#[derive(Serialize)]
struct OpenAIRequest {
    model: String,
    messages: Vec<OpenAIMessage>,
    temperature: f32,
    max_tokens: u32,
}

#[derive(Serialize)]
struct OpenAIMessage {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct OpenAIResponse {
    choices: Vec<OpenAIChoice>,
}

#[derive(Deserialize)]
struct OpenAIChoice {
    message: OpenAIMessageContent,
}

#[derive(Deserialize)]
struct OpenAIMessageContent {
    content: String,
}

impl OpenAIClient {
    /// Create a new OpenAI client.
    pub fn new(api_key: String) -> Self {
        Self {
            api_key,
            base_url: "https://api.openai.com/v1".to_string(),
            client: reqwest::Client::new(),
        }
    }

    /// Create with custom base URL (for proxies or compatible APIs).
    pub fn with_base_url(api_key: String, base_url: String) -> Self {
        Self {
            api_key,
            base_url,
            client: reqwest::Client::new(),
        }
    }
}

#[async_trait]
impl LLMClient for OpenAIClient {
    async fn complete(&self, prompt: &str, config: &LLMConfig) -> TradingResult<String> {
        let request = OpenAIRequest {
            model: config.model.clone(),
            messages: vec![OpenAIMessage {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
            temperature: config.temperature,
            max_tokens: config.max_tokens,
        };

        let response = self
            .client
            .post(format!("{}/chat/completions", self.base_url))
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .json(&request)
            .send()
            .await?;

        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(TradingError::LLMError(error_text));
        }

        let api_response: OpenAIResponse = response.json().await?;

        api_response
            .choices
            .first()
            .map(|c| c.message.content.clone())
            .ok_or_else(|| TradingError::LLMError("No response from API".to_string()))
    }

    fn provider(&self) -> &str {
        "openai"
    }
}

/// Anthropic Claude API client.
pub struct AnthropicClient {
    api_key: String,
    client: reqwest::Client,
}

#[derive(Serialize)]
struct AnthropicRequest {
    model: String,
    messages: Vec<AnthropicMessage>,
    max_tokens: u32,
}

#[derive(Serialize)]
struct AnthropicMessage {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct AnthropicResponse {
    content: Vec<AnthropicContent>,
}

#[derive(Deserialize)]
struct AnthropicContent {
    text: String,
}

impl AnthropicClient {
    /// Create a new Anthropic client.
    pub fn new(api_key: String) -> Self {
        Self {
            api_key,
            client: reqwest::Client::new(),
        }
    }
}

#[async_trait]
impl LLMClient for AnthropicClient {
    async fn complete(&self, prompt: &str, config: &LLMConfig) -> TradingResult<String> {
        let request = AnthropicRequest {
            model: config.model.clone(),
            messages: vec![AnthropicMessage {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
            max_tokens: config.max_tokens,
        };

        let response = self
            .client
            .post("https://api.anthropic.com/v1/messages")
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("Content-Type", "application/json")
            .json(&request)
            .send()
            .await?;

        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(TradingError::LLMError(error_text));
        }

        let api_response: AnthropicResponse = response.json().await?;

        api_response
            .content
            .first()
            .map(|c| c.text.clone())
            .ok_or_else(|| TradingError::LLMError("No response from API".to_string()))
    }

    fn provider(&self) -> &str {
        "anthropic"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_mock_client() {
        let client = MockLLMClient::new();
        let config = LLMConfig::default();

        let result = client.complete("test prompt", &config).await;
        assert!(result.is_ok());
    }

    #[tokio::test]
    async fn test_mock_client_with_responses() {
        let mut responses = HashMap::new();
        responses.insert(
            "earnings".to_string(),
            r#"{"sentiment": "POSITIVE", "confidence": 85}"#.to_string(),
        );

        let client = MockLLMClient::with_responses(responses);
        let config = LLMConfig::default();

        let result = client
            .complete("Apple reports strong earnings", &config)
            .await
            .unwrap();
        assert!(result.contains("POSITIVE"));
    }
}
